test/functional/plugin/lsp_spec.lua

```diff
@@ -3623,4 +3623,103 @@ describe('LSP', function()
       eq(expected, result)
     end)
   end)
+
+  describe('vim.lsp._watchfiles', function()
+    it('should send workspace/didChangeWatchedFiles when watched files change', function()
```

Comment:
At a high level, this test:
1. registers the capability dynamically, pointing at a temporary directory
2. generates some filesystem events
3. checks that the expected events are sent
4. re-registers to watch some different files
5. generates more filesystem events
6. checks that the expected events are received